### PR TITLE
[shopsys] added pagination to frontend api

### PR DESCRIPTION
**Pull request by DavidOstrozlik**
| Q             | A
| ------------- | ---
|Description, reason for the PR| products in fe api weren't paginated so developers had no chance of getting just a certain page of products and not all of them. This PR adds pagination
|New feature| Yes 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: packages/frontend-api/src/Resources/config/graphql-types/Pagination/EdgeInterface.types.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+EdgeInterface:
```
Comment (grossmannmartin) [inline packages/frontend-api/src/Resources/config/graphql-types/Pagination/EdgeInterface.types.yml]: Did you consider connection-relay? https://github.com/overblog/GraphQLBundle/blob/master/docs/definitions/relay/connection.md

File: project-base/src/Shopsys/ShopBundle/Resources/graphql-types/Pagination/ProductEdge.types.yml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+ProductEdge:
+    type: object
+    inherits:
+        - 'ProductEdgeDecorator'
+    config:
+        fields:
+            cursor:
```
Comment (pesektomas) [inline project-base/src/Shopsys/ShopBundle/Resources/graphql-types/Pagination/ProductEdge.types.yml]: Why is cursor field in this? Why cursor field not in decorator?

File: packages/frontend-api/src/Resources/config/graphql-types/Pagination/ConnectionDecorator.types.yml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+ConnectionDecorator:
+    type: object
```
Comment (pesektomas) [inline packages/frontend-api/src/Resources/config/graphql-types/Pagination/ConnectionDecorator.types.yml]: Why it is not relay-connection type? See this https://github.com/overblog/GraphQLBundle/blob/master/docs/definitions/relay/connection.md

File: project-base/src/Shopsys/ShopBundle/Resources/graphql-types/Pagination/Connection.types.yml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+Connection:
+    type: object
```
Comment (pesektomas) [inline project-base/src/Shopsys/ShopBundle/Resources/graphql-types/Pagination/Connection.types.yml]: Why it is not relay-connection type? See this https://github.com/overblog/GraphQLBundle/blob/master/docs/definitions/relay/connection.md

File: packages/frontend-api/src/Model/Resolver/Category/CategoryResolver.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+     */
+    public function resolver(string $uuid): Category
+    {
+        if (Uuid::isValid($uuid) === false) {
+            throw new UserError('Provided argument is not valid UUID.');
+        }
+
+        try {
+            $category = $this->categoryFacade->getByUuid($uuid);
+            return $category;
```
Comment (pesektomas) [inline packages/frontend-api/src/Model/Resolver/Category/CategoryResolver.php]:
```suggestion
            return $this->categoryFacade->getByUuid($uuid);;
```

File: packages/frontend-api/src/Model/Resolver/Pagination/NodeResolver.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+use Overblog\GraphQLBundle\Definition\Resolver\AliasedInterface;
+use Overblog\GraphQLBundle\Definition\Resolver\ResolverInterface;
+use Overblog\GraphQLBundle\Relay\Connection\Output\Edge;
+use Shopsys\ShopBundle\Model\Product\Product;
+
+class NodeResolver implements ResolverInterface, AliasedInterface
+{
+    /**
+     * @param \Overblog\GraphQLBundle\Relay\Connection\Output\Edge $edge
+     * @return mixed|Product
```
Comment (pesektomas) [inline packages/frontend-api/src/Model/Resolver/Pagination/NodeResolver.php]: You should use FQN. And remove Product from ShopBunde in use blocks :smile:

File: docs/frontend-api/pagination.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+{
+  products{}
+}
+```
+
+This query will return a `connection` object which consists of `pageInfo` and `edges`.
+
+`pageInfo` is object that represents information about current page of pagination that you are on.
+
+`edges` are array of objects that implements `EdgeInterface` and represent the products that you get.
```
Comment (pesektomas) [inline docs/frontend-api/pagination.md]: this part of documentation not longer true :smile:

File: packages/frontend-api/src/Resources/config/graphql-types/Pagination/EdgeInterface.types.yml (hunk 1 above)
Comment (DavidOstrozlik) [reply packages/frontend-api/src/Resources/config/graphql-types/Pagination/EdgeInterface.types.yml]: I changed it :slightly_smiling_face:

File: packages/frontend-api/src/Model/Resolver/Category/CategoryResolver.php (hunk 5 above)
Comment (DavidOstrozlik) [reply packages/frontend-api/src/Model/Resolver/Category/CategoryResolver.php]: fixed

File: packages/frontend-api/src/Model/Resolver/Pagination/NodeResolver.php (hunk 6 above)
Comment (DavidOstrozlik) [reply packages/frontend-api/src/Model/Resolver/Pagination/NodeResolver.php]: You are right. Fixed

File: docs/frontend-api/pagination.md (hunk 7 above)
Comment (pesektomas) [reply docs/frontend-api/pagination.md]: Can you please update all documentation? You have example with old ProductEdge.

File: packages/frontend-api/src/Model/Resolver/Pagination/NodeResolver.php (hunk 6 above)
Comment (henzigo) [reply packages/frontend-api/src/Model/Resolver/Pagination/NodeResolver.php]: Hello, there should not be a dependency on any entity (in this case on product entity). If you will add some blog pagination or orders pagination then you will always edit this line?

File: docs/frontend-api/pagination.md (hunk 7 above)
Comment (DavidOstrozlik) [reply docs/frontend-api/pagination.md]: @pesektomas

File: project-base/tests/FrontendApiBundle/Functional/Category/CategoryTest.php [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+                    }
+                }
+            }
+        ';
+
+        $arrayExpected = [
+            'data' => [
+                'category' => [
+                    'children' => [
+                        ['name' => t('TV, audio', [], 'dataFixtures', $this->domain->getDomainConfigById(Domain::FIRST_DOMAIN_ID)->getLocale())],
```
Comment (grossmannmartin) [inline project-base/tests/FrontendApiBundle/Functional/Category/CategoryTest.php]: can you extract locale into a separate variable?

File: packages/frontend-api/src/Model/Resolver/Products/ProductsResolver.php [hunk 9]
```diff
@@ -17,28 +23,107 @@ class ProductsResolver implements ResolverInterface, AliasedInterface
...
+        });
+
+        return $paginator->auto($argument, count($products));
+    }
+
+    /**
+     * @param \Shopsys\ShopBundle\Model\Category\Category $category
+     * @param int $offset
+     * @param int $limit
+     * @return array
```
Comment (grossmannmartin) [inline packages/frontend-api/src/Model/Resolver/Products/ProductsResolver.php]: Product[]

File: packages/frontend-api/src/Model/Resolver/Products/ProductsResolver.php [hunk 10]
```diff
@@ -17,28 +23,107 @@ class ProductsResolver implements ResolverInterface, AliasedInterface
...
+        $query = $queryBuilder->getQuery();
+        $query->setHint(Query::HINT_CUSTOM_OUTPUT_WALKER, SortableNullsWalker::class);
+
+        return $query->execute();
+    }
+
+    /**
+     * @param int $offset
+     * @param int $limit
+     * @return array
```
Comment (grossmannmartin) [inline packages/frontend-api/src/Model/Resolver/Products/ProductsResolver.php]: Product[]

File: packages/framework/src/Model/Product/ProductOnCurrentDomainFacade.php [hunk 11]
```diff
@@ -144,6 +146,39 @@ public function getPaginatedProductsInCategory(
...
+    public function getAllListableTranslatedAndOrderedQueryBuilder(
+        string $orderingModeId
+    ): QueryBuilder {
+        return $this->productRepository->getAllListableTranslatedAndOrderedQueryBuilder(
+            $this->domain->getId(),
+            $this->domain->getLocale(),
+            $orderingModeId,
+            $this->currentCustomer->getPricingGroup()
+        );
+    }
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Product/ProductOnCurrentDomainFacade.php]: can you please deprecate those methods?

**Pull request merged**